### PR TITLE
Feature/external event building

### PR DESCRIPTION
**Pull request by a-d-smith**
Please see these slides for a detailed explanation of the changes made
https://www.hep.phy.cam.ac.uk/~asmith/29062018-ExternalEventBuilding.pdf

If you want to run this with MicroBooNE (or anything that doesn't use the standard settings files), there are a few additional lines that are required for the pandora neutrino settings files:

```xml
<!-- Track and shower building -->
<algorithm type = "LArTrackParticleBuilding">
    <PfoListName>NeutrinoParticles3D</PfoListName>
    <VertexListName>NeutrinoVertices3D</VertexListName>
</algorithm>
<algorithm type = "LArPcaShowerParticleBuilding">
    <PfoListName>NeutrinoParticles3D</PfoListName>
    <VertexListName>NeutrinoVertices3D</VertexListName>
</algorithm>
```

### CODE REVIEW

File: larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+void LArPandoraExternalEventBuilding::CollectClearCosmicRays(const PFParticleToMetadata &particlesToMetadata, const PFParticleMap &particleMap, PFParticleVector &clearCosmics) const
+{
+    for (const auto &entry : particlesToMetadata)
+    {
+        try
+        {
+            // Check if the parent PFParticle is clear cosmic
+            const auto parentIterator(particlesToMetadata.find(LArPandoraHelper::GetParentPFParticle(particleMap, entry.first)));
+
+            if (static_cast<bool>(std::round(this->GetMetadataValue(parentIterator->second, "IsClearCosmic"))))
```
Comment (PandoraPFA) [inline larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc]: Great use of `std::round`!

File: larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void LArPandoraExternalEventBuilding::CollectSlices(const PFParticleToMetadata &particlesToMetadata, const PFParticleMap &particleMap, SliceVector &slices) const
+{
+    std::map<unsigned int, float> nuScores;
+    std::map<unsigned int, PFParticleVector> crHypotheses;
+    std::map<unsigned int, PFParticleVector> nuHypotheses;
+
+    // Collect the slice information
+    for (const auto &entry : particlesToMetadata)
```
Comment (PandoraPFA) [inline larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc]: Again check ordering is defined and reproducible.

File: larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    {
+        // Find the parent PFParticle
+        const auto parentIterator(particlesToMetadata.find(LArPandoraHelper::GetParentPFParticle(particleMap, entry.first)));
+        if (parentIterator == particlesToMetadata.end())
+            throw cet::exception("LArPandoraExternalEventBuilding") << "Can't find the parent of input PFParticle" << std::endl;
+        
+        const int pdg(std::abs(parentIterator->first->PdgCode())); 
+        const bool isNeutrino(std::abs(pdg) == pandora::NU_E || std::abs(pdg) == pandora::NU_MU || std::abs(pdg) == pandora::NU_TAU);
```
Comment (PandoraPFA) [inline larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc]: Check/use/update `LArPandoraHelper::IsNeutrino`

File: larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+        {
+            sliceId = static_cast<unsigned int>(std::round(this->GetMetadataValue(parentIterator->second, "SliceIndex")));
+            nuScore = this->GetMetadataValue(parentIterator->second, "NuScore");
+        }
+        catch (const cet::exception &exception)
+        {
+            // The above information should only unavailable for clear cosmic PFParticles
+            try
+            {
+                if (static_cast<bool>(std::round(this->GetMetadataValue(parentIterator->second, "IsClearCosmic")))) continue;
```
Comment (PandoraPFA) [inline larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc]: Maybe better to fully-resolve fallout from first exception before any further operations? Then move on to this treatment? I know you raise the same exception instance again, but don't much like much work inside a catch block.

Also , prefer not to have continue on the same line, but hey...

File: larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+            nuHypotheses[sliceId].push_back(entry.first);
+        }
+        else 
+        {
+            crHypotheses[sliceId].push_back(entry.first);
+        }
+    }
+
+    // Produce the slices
+    for (unsigned int sliceId = 1; sliceId <= nuScores.size(); ++sliceId)
```
Comment (PandoraPFA) [inline larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc]: Here we see the "inside knowledge" that sliceIds start at 1... Might want to add an ATTN or similar.

File: larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+    collectedParticles.insert(collectedParticles.end(), clearCosmics.begin(), clearCosmics.end());
+
+    for (const auto &slice : slices)
+    {
+        const PFParticleVector &particles(slice.IsTaggedAsNeutrino() ? slice.GetNeutrinoHypothesis() : slice.GetCosmicRayHypothesis());
+        collectedParticles.insert(collectedParticles.end(), particles.begin(), particles.end());
+    }
+
+    // ATTN This is required to ensure that the particles remain in order!
+    for (const auto &part : allParticles)
```
Comment (PandoraPFA) [inline larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc]: Maybe one more line to elaborate upon need for this?

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx [hunk 7]
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
...
+    std::unique_ptr< art::Assns<recob::PFParticle, recob::SpacePoint> >                 outputParticlesToSpacePoints( new art::Assns<recob::PFParticle, recob::SpacePoint> );
+    std::unique_ptr< art::Assns<recob::PFParticle, recob::Cluster> >                    outputParticlesToClusters( new art::Assns<recob::PFParticle, recob::Cluster> );
+    std::unique_ptr< art::Assns<recob::PFParticle, recob::Vertex> >                     outputParticlesToVertices( new art::Assns<recob::PFParticle, recob::Vertex> );
+    std::unique_ptr< art::Assns<recob::PFParticle, anab::T0> >                          outputParticlesToT0s( new art::Assns<recob::PFParticle, anab::T0> );
+    std::unique_ptr< art::Assns<recob::Cluster, recob::Hit> >                           outputClustersToHits( new art::Assns<recob::Cluster, recob::Hit> );
+    std::unique_ptr< art::Assns<recob::SpacePoint, recob::Hit> >                        outputSpacePointsToHits( new art::Assns<recob::SpacePoint, recob::Hit> );
 
-    if (!settings.m_pPrimaryPandora)
-        throw cet::exception("LArPandora") << " LArPandoraOutput::ProduceArtOutput --- primary Pandora instance does not exist ";
+    // ---
```
Comment (PandoraPFA) [inline larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: If you need to break up implementation with ascii art, then refactor... :)

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx [hunk 8]
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
...
 
-    if (!settings.m_pProducer)
-        throw cet::exception("LArPandora") << " LArPandoraOutput::ProduceArtOutput --- pointer to ART Producer module does not exist ";
+    // Collect immutable lists of pandora collections that we should convert to ART format
+    // The index of an object in a list will become the ID of the ART object produced
+    // Also obtain mappings between IDs in different collections which will become associations
```
Comment (PandoraPFA) [inline larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: This information would normally go in the header file, with the function prototype/definition.

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx [hunk 9]
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
...
+    const pandora::CaloHitList threeDHitList(LArPandoraOutput::Collect3DHits(pfoList, pfoToThreeDHitsMap));
 
-    // Set up ART outputs from RecoBase, AnalysisBase and LArPandoraObjects
-    std::unique_ptr< std::vector<recob::PFParticle> >                 outputParticles( new std::vector<recob::PFParticle> );
-    std::unique_ptr< std::vector<recob::SpacePoint> >                 outputSpacePoints( new std::vector<recob::SpacePoint> );
-    std::unique_ptr< std::vector<recob::Cluster> >                    outputClusters( new std::vector<recob::Cluster> );
-    std::unique_ptr< std::vector<recob::Vertex> >                     outputVertices( new std::vector<recob::Vertex> );
-    std::unique_ptr< std::vector<anab::T0> >                          outputT0s( new std::vector<anab::T0> );
-    std::unique_ptr< std::vector<larpandoraobj::PFParticleMetadata> > outputParticleMetadata( new std::vector<larpandoraobj::PFParticleMetadata> );
+    // ---
```
Comment (PandoraPFA) [inline larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: Actually not looking too long, although beyond the 7+/-2 lines we should aspire too!!!

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx [hunk 10]
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
...
+    // Identify the pandora worker instances by their name
+    const pandora::Pandora *pSlicingWorker(nullptr);
+    const pandora::Pandora *pSliceNuWorker(nullptr);
+    const pandora::Pandora *pSliceCRWorker(nullptr);
 
-    for (const pandora::ParticleFlowObject *const pPfo : pfoVector)
+    for (const pandora::Pandora *const pPandora : MultiPandoraApi::GetDaughterPandoraInstanceList(pMasterPandora))
     {
-        particleMap.insert( std::pair<const pandora::ParticleFlowObject*, size_t>(pPfo, particleCounter++) );
+        const std::string name(pPandora->GetName());
```
Comment (PandoraPFA) [inline larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: const reference.

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx [hunk 11]
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
...
+    const pandora::PfoList *pParentPfoList(nullptr);
+    PANDORA_THROW_RESULT_IF(pandora::STATUS_CODE_SUCCESS, !=, PandoraApi::GetCurrentPfoList(*pMasterPandora, pParentPfoList));
 
-    // Build maps of pandora::Pfos and build recob::vertices
-    ThreeDParticleMap particleMap;
-    ThreeDVertexMap vertexMap;
+    // Identify the pandora worker instances by their name
+    const pandora::Pandora *pSlicingWorker(nullptr);
+    const pandora::Pandora *pSliceNuWorker(nullptr);
+    const pandora::Pandora *pSliceCRWorker(nullptr);
```
Comment (PandoraPFA) [inline larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: Obviously need to be super careful here for simple use cases where there are no slice workers, etc. and these are left holding `nullptr` values.

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx [hunk 12]
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
...
     }
 
-    // Loop over pandora::Pfos and build recob::PFParticles
-    for (const pandora::ParticleFlowObject *const pPfo : pfoVector)
-    {
-        // Get Pfo ID
-        ThreeDParticleMap::const_iterator iter = particleMap.find(pPfo);
-        if (particleMap.end() == iter)
-            throw cet::exception("LArPandora") << " LArPandoraOutput::ProduceArtOutput --- found an unassociated particle";
+    if (!pSlicingWorker || !pSliceNuWorker || !pSliceCRWorker)
```
Comment (PandoraPFA) [inline larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: What about other use cases? Do we just rely on setting `enableAllOutcomes` to false there, or else an exception will be raised quite quickly and terminate the job? Not a bad approach...

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx [hunk 13]
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
...
+        throw cet::exception("LArPandora") << " LArPandoraOutput::CollectAllPfoOutcomes--- Can't produce all outcomes for a non-consolidated pandora producer ";
+
+    // Collect slices under bothe reconstruction outcomes
+    const pandora::PfoList *pSlicePfoList(nullptr);
+    PANDORA_THROW_RESULT_IF(pandora::STATUS_CODE_SUCCESS, !=, PandoraApi::GetCurrentPfoList(*pSlicingWorker, pSlicePfoList));
 
-        const size_t pfoIdCode(iter->second);
+    for (unsigned int sliceIndex = 0; sliceIndex < pSlicePfoList->size(); ++sliceIndex)
```
Comment (PandoraPFA) [inline larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: Here sliceIndex starts at 0. Are we being consistent throughout?

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx [hunk 14]
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
...
+    for (const pandora::ParticleFlowObject *const pPfo : *pParentPfoList)
+    {
+        bool isClearCosmic(false);
+        try
         {
-            if (parentList.size() != 1)
-                throw cet::exception("LArPandora") << " LArPandoraOutput::ProduceArtOutput --- this particle has multiple parent particles ";
+            isClearCosmic = pPfo->GetPropertiesMap().at("IsClearCosmic");
+        }
+        catch (...)
```
Comment (PandoraPFA) [inline larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: Does this need to be `catch (...)`? Instances of which types may be raised?

File: larpandora/LArPandoraObjects/classes_def.xml [hunk 15]
```diff
@@ -10,4 +10,10 @@
             
     <class name="art::Assns<larpandoraobj::PFParticleMetadata, recob::PFParticle,void>"/>
     <class name="art::Wrapper<art::Assns<larpandoraobj::PFParticleMetadata, recob::PFParticle,void> >"/>
+            
+    <class name="art::Assns<recob::PFParticle,bool,void>"/>
+    <class name="art::Wrapper<art::Assns<recob::PFParticle,bool,void> >"/>
+            
+    <class name="art::Assns<bool,recob::PFParticle,void>"/>
+    <class name="art::Wrapper<art::Assns<bool,recob::PFParticle,void> >"/>
```
[the remark this answers is not in the document]
Comment (a-d-smith) [reply larpandora/LArPandoraObjects/classes_def.xml]: Sorry, this was something I forgot to remove from an old iteration!

File: larpandora/LArPandoraInterface/LArPandoraOutput.h [hunk 16]
```diff
@@ -7,15 +7,20 @@
...
+#include "lardata/Utilities/AssociationUtil.h"
+
 #include "lardataobj/RecoBase/Cluster.h"
 
 #include "larreco/RecoAlg/ClusterRecoUtil/ClusterParamsAlgBase.h"
 
 #include "larpandora/LArPandoraInterface/ILArPandora.h"
 #include "larpandora/LArPandoraInterface/LArPandoraHelper.h"
 
+#include "Pandora/PandoraInternal.h"
```
[the remark this answers is not in the document]
Comment (a-d-smith) [reply larpandora/LArPandoraInterface/LArPandoraOutput.h]: I used this to pick-up all the typedefs for `ClusterList`, `CaloHitVector`, etc... If preferable I can just make these typedefs here?

File: larpandora/LArPandoraInterface/LArPandoraOutput.h [hunk 17]
```diff
@@ -50,46 +65,384 @@ class LArPandoraOutput
...
+    /**
+     *  @brief  Add an association between objects with two given ids
+     *
+     *  @param  event the ART event
+     *  @param  pProducer the address of the producer module
+     *  @param  idA the id of an object of type A
+     *  @param  idB the id of an object of type B to associate to the first object
+     *  @param  association the output association to update
+     */
+    template <typename A, typename B>
```
[the remark this answers is not in the document]
Comment (a-d-smith) [reply larpandora/LArPandoraInterface/LArPandoraOutput.h]: I guess this links up with your above comment about including `PandoraInternal.h`. Would you rather I didn't include that file?

File: larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+    
+    LArPandoraExternalEventBuilding(LArPandoraExternalEventBuilding const &) = delete;
+    LArPandoraExternalEventBuilding(LArPandoraExternalEventBuilding &&) = delete;
+    LArPandoraExternalEventBuilding & operator = (LArPandoraExternalEventBuilding const &) = delete;
+    LArPandoraExternalEventBuilding & operator = (LArPandoraExternalEventBuilding &&) = delete;
+
+    void produce(art::Event &evt) override;
+
+private:
+    typedef std::map<art::Ptr<recob::PFParticle>, art::Ptr<larpandoraobj::PFParticleMetadata> > PFParticleToMetadata;
```
[the remark this answers is not in the document]
Comment (a-d-smith) [reply larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc]: This is a new typedef, the nearest typedef in LArPandoraHelper is PFParticle to vector<Metadata>

File: larpandora/LArPandoraInterface/LArPandoraOutput.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (PandoraPFA) [reply larpandora/LArPandoraInterface/LArPandoraOutput.h]: Well, a strict include "what you use" policy means you should include this file, but I tend to recommend that people just include Api/PandoraApi.h and it should give them everything they need to write a Pandora application.

This starts to get subjective, and there are arguments for and against. Ultimately include what you use would probably win, but just including the Api header would be clean.

File: larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+
+namespace lar_pandora
+{
+
+LArPandoraExternalEventBuilding::LArPandoraExternalEventBuilding(fhicl::ParameterSet const &pset) :
+    m_inputProducerLabel(pset.get<std::string>("InputProducerLabel")),
+    m_trackProducerLabel(pset.get<std::string>("TrackProducerLabel")),
+    m_showerProducerLabel(pset.get<std::string>("ShowerProducerLabel")),
+    m_hitProducerLabel(pset.get<std::string>("HitProducerLabel")),
+    m_shouldProduceT0s(pset.get<bool>("ShouldProduceT0s", false)),
```
[the remark this answers is not in the document]
Comment (a-d-smith) [reply larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc]: Looking it up, LArPandora actually has no default for ShouldRunStitching (which controls if T0s are produced). I will remove this default to be consistent

File: larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+     *  @return the value in the metadata corresponding to the input key
+     */
+    float GetMetadataValue(const art::Ptr<larpandoraobj::PFParticleMetadata> &metadata, const std::string &key) const;
+
+    std::string                         m_inputProducerLabel;  ///< Label for the Pandora instance that produced the collections we want to consolidated
+    std::string                         m_trackProducerLabel;  ///< Label for the track producer using the Pandora instance that produced the collections we want to consolidate
+    std::string                         m_showerProducerLabel; ///< Label for the shower producer using the Pandora instance that produced the collections we want to consolidate
+    std::string                         m_hitProducerLabel;    ///< Label for the hit producer that was used as input to the Pandora instance specified
+    bool                                m_shouldProduceT0s;    ///< If we should produce T0s (relevant when stitching over multiple drift volumes)
+    art::InputTag                       m_pandoraTag;          ///< The input tag for the pandora producer
```
[the remark this answers is not in the document]
Comment (a-d-smith) [reply larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc]: Me neither until recently, they are nifty!

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx (hunk 8 above)
Comment (a-d-smith) [reply larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: This information relates to all of the `Collect*` functions, so I wasn't sure where to put it without just repeating.

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx (hunk 12 above)
Comment (a-d-smith) [reply larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: Yes, exactly that's the idea.

File: larpandora/LArPandoraInterface/LArPandoraOutput.cxx (hunk 13 above)
Comment (a-d-smith) [reply larpandora/LArPandoraInterface/LArPandoraOutput.cxx]: The PFO collection names start from zero, but the slices (as logged by Pandora) start from 1. Here I am just using the numbering system in each case from Pandora, but I can adopt one or the other if you prefer?

**Comment by a-d-smith** [comment]
Hello! 

I have hopefully addressed your comments so far as best I can. I have also - where easy to do so - used the LArPfoHelper instead of reinventing the wheel, and simplified some confusing try-catch logic. There are still some nastinesses lurking where I haven't changed the implementation of certain functions from the pre-factored code.

I will circulate some slides soon with instructions on how best to test. I think the event dumping module I wrote a while ago will be the most useful tool as it can be easily diffed! In my test so far, things seem to be working as expected but (as you say) much more testing is required for all the experiments this touches.

**Comment by PandoraPFA** [comment]
Hi Andy,

Many thanks!

I think you should now distribute some simple slides with some clear tests indicated - ideally a concrete before/after here is what should/shouldn't change. Then you'll need to gently persuade :) everyone to do a bit of testing - I should have time too, especially if the instructions about e.g. the .fcl configuration for test jobs are nice and clear.

Once we're happy with that, I suggest that you and I sit down and read through LArPandoraOutput.cxx together again to double check, then we go for a talk and a release. Does that work for you?

John

**Comment by StevenGreen1** [comment]
Merged via the command line.

**Pull request closed without merging**